# APSEV Data Cleaning Notebook

## **Note:** This was completed on a Windows Machine running Windows 10 64bit.
Code has been tailored to run universally on other operating systems however this is not guarenteed. If you encounter an error please report so it can be investigated and rectified. 

In [ ]:
# Let's import our dependancies for the cleaning process
import pandas as pd
import os
import numpy as np

# Making pandas display all columns
pd.options.display.max_columns = None

In [ ]:
# Just to verify we can make sure we verify our working directory
os.getcwd()

From here we need to import the dataset but there is an issue:  the dataset is presented in an xlsx format, and to properly interogate it we will need to convert it to csv.  

Fortunately pandas is able to do this natively with the *pandas.DataFrame.to_csv* function. 

In [3]:
data_xls = pd.read_excel('../dataset/school-profile-2008-2019-ACARA.xlsx', 'SchoolProfile 2008-2019', index_col=None) # We use the SchoolProfile 2008-2019 arguement to ensure we pick the correct sheet
data_xls.to_csv('school_profile_2008-2019.csv', encoding='utf-8', index=False)

In [4]:
# We will now verify the CSV has been created successfully by checking it. 

os.listdir(os.getcwd())

['.ipynb_checkpoints',
 'APSEV_data_cleaning.ipynb',
 'school_profile_2008-2019.csv']

VOLA! The dataset is now converted into a CSV file format ready to be cleaned, so lets get it into a dataframe!

In [5]:
# Importing the CSV into a dataframe
df1_csv = pd.read_csv('../Code/school_profile_2008-2019.csv')

In [6]:
# Checking the dataset imported correctly
df1_csv.head()

Calendar Year  ACARA SML ID  AGE ID                     School Name  \
0           2008         40000     3.0  Corpus Christi Catholic School   
1           2009         40000     3.0  Corpus Christi Catholic School   
2           2010         40000     3.0  Corpus Christi Catholic School   
3           2011         40000     3.0  Corpus Christi Catholic School   
4           2012         40000     3.0  Corpus Christi Catholic School   

      Suburb State  Postcode School Sector School Type           Campus Type  \
0  Bellerive   TAS      7018      Catholic     Primary  School Single Entity   
1  Bellerive   TAS      7018      Catholic     Primary  School Single Entity   
2  Bellerive   TAS      7018      Catholic     Primary  School Single Entity   
3  Bellerive   TAS      7018      Catholic     Primary  School Single Entity   
4  Bellerive   TAS      7018      Catholic     Primary  School Single Entity   

  Rolled Reporting Description                           School URL  \
0         Individual Reporting  http://www.corpuschristi.tas.edu.au   
1         Individual Reporting  http://www.corpuschristi.tas.edu.au   
2         Individual Reporting  http://www.corpuschristi.tas.edu.au   
3         Individual Reporting  http://www.corpuschristi.tas.edu.au   
4         Individual Reporting  http://www.corpuschristi.tas.edu.au   

                        Governing Body              Governing Body URL  \
0  Tasmanian Catholic Education Office  http://www.catholic.tas.edu.au   
1  Tasmanian Catholic Education Office  http://www.catholic.tas.edu.au   
2  Tasmanian Catholic Education Office  http://www.catholic.tas.edu.au   
3  Tasmanian Catholic Education Office  http://www.catholic.tas.edu.au   
4  Tasmanian Catholic Education Office  http://www.catholic.tas.edu.au   

  Year Range     Geolocation   ICSEA  ICSEA Percentile  \
0  U, Prep-6  Inner Regional  1055.0              75.0   
1  U, Prep-6  Inner Regional  1055.0              75.0   
2  U, Prep-6  Inner Regional  1044.0              70.0   
3  U, Prep-6  Inner Regional  1046.0              71.0   
4  U, Prep-6  Inner Regional  1041.0              69.0   

   Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
0                    13.0                          19.0   
1                    13.0                          19.0   
2                    28.0                          19.0   
3                    19.0                          30.0   
4                     9.0                          35.0   

   Upper Middle SEA Quarter (%)  Top SEA Quarter (%)  Teaching Staff  \
0                          37.0                 31.0            29.0   
1                          37.0                 31.0            27.0   
2                          30.0                 23.0            26.0   
3                          25.0                 25.0            25.0   
4                          28.0                 28.0            24.0   

   Full Time Equivalent Teaching Staff  Non-Teaching Staff  \
0                                 19.0                12.0   
1                                 19.1                15.0   
2                                 18.6                18.0   
3                                 18.1                16.0   
4                                 18.1                16.0   

   Full Time Equivalent Non-Teaching Staff  Total Enrolments  \
0                                      6.9             410.0   
1                                      8.7             409.0   
2                                     10.2             392.0   
3                                      9.0             385.0   
4                                      9.4             391.0   

   Girls Enrolments  Boys Enrolments  Full Time Equivalent Enrolments  \
0             185.0            225.0                            410.0   
1             184.0            225.0                            409.0   
2             173.0            219.0                            392.0   
3             173.

And it worked! now to see how much data we are working with using the .shape function

In [7]:
# Checking the shape of the dataframe
df1_csv.shape

(111519, 32)

That is alot of data! lets start by exploring our metrics

In [8]:
len(df1_csv.columns), df1_csv.columns

(32,
 Index(['Calendar Year', 'ACARA SML ID', 'AGE ID', 'School Name', 'Suburb',
        'State', 'Postcode', 'School Sector', 'School Type', 'Campus Type',
        'Rolled Reporting Description', 'School URL', 'Governing Body',
        'Governing Body URL', 'Year Range', 'Geolocation', 'ICSEA',
        'ICSEA Percentile', 'Bottom SEA Quarter (%)',
        'Lower Middle SEA Quarter (%)', 'Upper Middle SEA Quarter (%)',
        'Top SEA Quarter (%)', 'Teaching Staff',
        'Full Time Equivalent Teaching Staff', 'Non-Teaching Staff',
        'Full Time Equivalent Non-Teaching Staff', 'Total Enrolments',
        'Girls Enrolments', 'Boys Enrolments',
        'Full Time Equivalent Enrolments', 'Indigenous Enrolments (%)',
        'Language Background Other Than English (%)'],
       dtype='object'))

The 32 columns check out with the XLSX that we inspected after downloading from the source, there are a number of columsn that will not be useful for us here such as the 'ACARA SML ID' and 'AGE ID' colummns.  We will check them out just in case.

In [9]:
# The following will check the frequencey of unique values in the ACARA SML ID and AGE ID columns

id_frequency = df1_csv.groupby(['ACARA SML ID', 'AGE ID'])['School Name'].count()
id_frequency

ACARA SML ID  AGE ID 
40000         3.0        12
40001         4.0        12
40002         5.0        12
40003         7.0        12
40004         9.0        12
                         ..
52745         85821.0     1
52747         85759.0     1
52756         85689.0     1
52784         85719.0     1
52785         85721.0     1
Name: School Name, Length: 10838, dtype: int64

We can see that the above indicates that these are just values assigned by ACARA to the schools, same with the AGE ID.  Officially the documentation for the dataset dictates the following: 

***ACARA SML ID:*** *Unique ID allocated to a given school by ACARA*

***AGE ID:*** *The Australian Government Department of Education ID (formerly DEEWR_ID).*

These appear to be internal use identifying numbers and we do not have a need for them in our use case.  Time to get rid of them.

Another deep dive into the dataset and we can see that there are more columns that do not pertain to us, these are:

- Rolled Reporting Description: Does not matter to our use case, this is only for noting how the data was obtained / reported
- Full Time Equivalent Enrolments: This is to drill down students that are less than 5 days a week, as there is no further data for this metric we will lable it as clutter and remove it. 
- Full Time Equivalent Teaching staff / Full Time Equvalent Non-teaching Staff:  These do the same as above and relate to staff members that are on split days, employed less than 5 days a week etc.  Again for our use case this level of detail is not needed and thus we will remove them. 

In [10]:
# We will start by allocating the columns to drop to a variable

cols_to_drop = df1_csv[['ACARA SML ID', 'AGE ID', 'Rolled Reporting Description', 'Full Time Equivalent Teaching Staff', 'Full Time Equivalent Non-Teaching Staff', 'Full Time Equivalent Enrolments']]

cols_to_drop

ACARA SML ID   AGE ID Rolled Reporting Description  \
0              40000      3.0         Individual Reporting   
1              40000      3.0         Individual Reporting   
2              40000      3.0         Individual Reporting   
3              40000      3.0         Individual Reporting   
4              40000      3.0         Individual Reporting   
...              ...      ...                          ...   
111514         52747  85759.0         Individual Reporting   
111515         52756  85689.0         Individual Reporting   
111516         52784  85719.0         Individual Reporting   
111517         52785  85721.0         Individual Reporting   
111518         52792      NaN         Individual Reporting   

        Full Time Equivalent Teaching Staff  \
0                                      19.0   
1                                      19.1   
2                                      18.6   
3                                      18.1   
4                                      18.1   
...                                     ...   
111514                                  4.3   
111515                                 12.5   
111516                                  6.4   
111517                                  5.8   
111518                                  4.5   

        Full Time Equivalent Non-Teaching Staff  \
0                                           6.9   
1                                           8.7   
2                                          10.2   
3                                           9.0   
4                                           9.4   
...                                         ...   
111514                                      0.4   
111515                                      4.0   
111516                                      2.9   
111517                                      1.8   
111518                                      5.8   

        Full Time Equivalent Enrolments  
0                                 410.0  
1                                 409.0  
2                                 392.0  
3                                 385.0  
4                                 391.0  
...                                 ...  
111514                             16.0  
111515                            217.0  
111516                            105.0  
111517                             63.0  
111518                             49.0  

[111519 rows x 6 columns]

In [11]:
# Now lets get rid of them and check they are gone
df2_csv = df1_csv.drop(cols_to_drop, axis=1)

df2_csv.head()

Calendar Year                     School Name     Suburb State  Postcode  \
0           2008  Corpus Christi Catholic School  Bellerive   TAS      7018   
1           2009  Corpus Christi Catholic School  Bellerive   TAS      7018   
2           2010  Corpus Christi Catholic School  Bellerive   TAS      7018   
3           2011  Corpus Christi Catholic School  Bellerive   TAS      7018   
4           2012  Corpus Christi Catholic School  Bellerive   TAS      7018   

  School Sector School Type           Campus Type  \
0      Catholic     Primary  School Single Entity   
1      Catholic     Primary  School Single Entity   
2      Catholic     Primary  School Single Entity   
3      Catholic     Primary  School Single Entity   
4      Catholic     Primary  School Single Entity   

                            School URL                       Governing Body  \
0  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
1  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
2  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
3  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
4  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   

               Governing Body URL Year Range     Geolocation   ICSEA  \
0  http://www.catholic.tas.edu.au  U, Prep-6  Inner Regional  1055.0   
1  http://www.catholic.tas.edu.au  U, Prep-6  Inner Regional  1055.0   
2  http://www.catholic.tas.edu.au  U, Prep-6  Inner Regional  1044.0   
3  http://www.catholic.tas.edu.au  U, Prep-6  Inner Regional  1046.0   
4  http://www.catholic.tas.edu.au  U, Prep-6  Inner Regional  1041.0   

   ICSEA Percentile  Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
0              75.0                    13.0                          19.0   
1              75.0                    13.0                          19.0   
2              70.0                    28.0                          19.0   
3              71.0                    19.0                          30.0   
4              69.0                     9.0                          35.0   

   Upper Middle SEA Quarter (%)  Top SEA Quarter (%)  Teaching Staff  \
0                          37.0                 31.0            29.0   
1                          37.0                 31.0            27.0   
2                          30.0                 23.0            26.0   
3                          25.0                 25.0            25.0   
4                          28.0                 28.0            24.0   

   Non-Teaching Staff  Total Enrolments  Girls Enrolments  Boys Enrolments  \
0                12.0             410.0             185.0            225.0   
1                15.0             409.0             184.0            225.0   
2                18.0             392.0             173.0            219.0   
3                16.0             385.0             173.0            212.0   
4                16.0             391.0             185.0            206.0   

   Indigenous Enrolments (%)  Language Background Other Than English (%)  
0                        1.0                                         4.0  
1                        2.0                                         4.0  
2                        2.0                                         4.0  
3                        0.0                                         3.0  
4                        1.0                                         3.0

We are feeling cleaner already!  32 columns down to 26, it is now time to look for missing data

In [12]:
# Checking for missing values

df2_csv.isna().sum()

Calendar Year                                     0
School Name                                       0
Suburb                                            0
State                                             0
Postcode                                          0
School Sector                                     0
School Type                                       0
Campus Type                                       0
School URL                                      345
Governing Body                                  101
Governing Body URL                              101
Year Range                                      631
Geolocation                                       0
ICSEA                                          6060
ICSEA Percentile                               6578
Bottom SEA Quarter (%)                        10258
Lower Middle SEA Quarter (%)                  10258
Upper Middle SEA Quarter (%)                  10258
Top SEA Quarter (%)                           10258
Teaching Sta

As we can see in the above, there are a number of metrics, and a number of missing values, but not all is what it seems.  

As an example an all girls school would not report on male student enrolment and thus it will be missing a value for that segment of data. 

Other values may also be missing based on not being reported, for an example there may be schools with no other language backgrounds other than english and therefore they may not report on it at all, instead of populating it with 0%

We will start by converting the missing values to a percentage so we can get a better idea of the overall picture. 

In [13]:
# Get the raw count of missing values
missing_values = df2_csv.isna().sum()

# Divide it by the total size and multiply by 100 to get the percent of missing values

missing_percent = ((missing_values / df2_csv.shape[0]) * 100)
missing_percent

Calendar Year                                  0.000000
School Name                                    0.000000
Suburb                                         0.000000
State                                          0.000000
Postcode                                       0.000000
School Sector                                  0.000000
School Type                                    0.000000
Campus Type                                    0.000000
School URL                                     0.309364
Governing Body                                 0.090568
Governing Body URL                             0.090568
Year Range                                     0.565823
Geolocation                                    0.000000
ICSEA                                          5.434052
ICSEA Percentile                               5.898546
Bottom SEA Quarter (%)                         9.198433
Lower Middle SEA Quarter (%)                   9.198433
Upper Middle SEA Quarter (%)                   9

As we can see, the impact the missing values have on the overall dataframe are pretty tiny, however as small as they are some of these values are still very important, others not so much. 

Lets create a list of columns to fix

In [14]:
columns_to_fix = list(missing_percent[(missing_percent > 0)].index)
columns_to_fix

['School URL',
 'Governing Body',
 'Governing Body URL',
 'Year Range',
 'ICSEA',
 'ICSEA Percentile',
 'Bottom SEA Quarter (%)',
 'Lower Middle SEA Quarter (%)',
 'Upper Middle SEA Quarter (%)',
 'Top SEA Quarter (%)',
 'Teaching Staff',
 'Non-Teaching Staff',
 'Total Enrolments',
 'Girls Enrolments',
 'Boys Enrolments',
 'Indigenous Enrolments (%)',
 'Language Background Other Than English (%)']

We will gather columns with non numerical values (such as URLs etc) so we can better fix the missing data. 

In [15]:
# Creating variable of non numerical columns
non_numerical_variables = ['School URL', 'Governing Body', 'Governing Body URL']


In [16]:
# Filling the missing values in the non numerical columns with 'unknown'
unknown_condition = {col:'unknown' for col in non_numerical_variables}
unknown_condition

{'School URL': 'unknown',
 'Governing Body': 'unknown',
 'Governing Body URL': 'unknown'}

In [17]:
# We will fill the missing values with unknown in our current dataframe we are working from
df2_csv = df2_csv.fillna(unknown_condition)

In [18]:
#Checking the NaNs were filled
df2_csv.isna().sum()

Calendar Year                                     0
School Name                                       0
Suburb                                            0
State                                             0
Postcode                                          0
School Sector                                     0
School Type                                       0
Campus Type                                       0
School URL                                        0
Governing Body                                    0
Governing Body URL                                0
Year Range                                      631
Geolocation                                       0
ICSEA                                          6060
ICSEA Percentile                               6578
Bottom SEA Quarter (%)                        10258
Lower Middle SEA Quarter (%)                  10258
Upper Middle SEA Quarter (%)                  10258
Top SEA Quarter (%)                           10258
Teaching Sta

Success!!

Lets now move onto the numerical values, oh but first we have an anomoly!

the column 'Year Range' is responsible for showing if a school teaches prep to 6, 7 to 12 and a few other variations. 

As per the dataset information:

***The range of year levels offered by the school. It is a derived field based on the year levels offered for the current year.***
- a) Year 1 minus 2: P = Preschool, K = Kindergarten.
- b) Year 1 minus 1: K = Kindergarten, Prep - Preparatory, R = Reception, PP = Pre-Primary, T=Transition.
- c) Years 1 - 12.
- d) U - Ungraded.

In [19]:
# Confirming the entries
df2_csv['Year Range'].head()

0    U, Prep-6
1    U, Prep-6
2    U, Prep-6
3    U, Prep-6
4    U, Prep-6
Name: Year Range, dtype: object

This is getting a touch complex, but as the grades a school teaches is very important we will need to deal with this column.  First any school that does not report what year levels it teaches we will need to check out why, and decide wether we need to keep these entries or not. 

In [20]:
# We will aggrigate the missing values into their own df so we can investigate why these values might be missing
year_range_missing = df2_csv[df2_csv['Year Range'].isna()]
year_range_missing.head()

Calendar Year                                   School Name  \
2736           2008           Saint Mary MacKillop College Albury   
3626           2010                             Elizabeth College   
4117           2010                            Kioma State School   
4295           2008               Hallam Senior Secondary College   
4367           2008  Field of Mars Environmental Education Centre   

            Suburb State  Postcode School Sector School Type  \
2736       Jindera   NSW      2642   Independent    Combined   
3626  North Hobart   TAS      7000    Government   Secondary   
4117         Kioma   QLD      4498    Government     Primary   
4295        Hallam   VIC      3803    Government   Secondary   
4367     East Ryde   NSW      2113    Government     Special   

               Campus Type                                         School URL  \
2736  School Single Entity                            http://smmc.nsw.edu.au/   
3626  School Single Entity  http://education.tas.edu.au/academy/elizabeth/...   
4117  School Single Entity                       http://www.kiomass.eq.edu.au   
4295    School Head Campus                    http://www.hallamssc.vic.edu.au   
4367  School Single Entity        http://www.fieldofmar-e.schools.nsw.edu.au/   

                                         Governing Body  \
2736  Association of Independent Schools of New Sout...   
3626                   Department of Education Tasmania   
4117                 Department of Education Queensland   
4295      Department of Education and Training Victoria   
4367                        Department of Education NSW   

                   Governing Body URL Year Range     Geolocation   ICSEA  \
2736         http://www.aisnsw.edu.au        NaN  Inner Regional     NaN   
3626  http://www.education.tas.gov.au        NaN  Inner Regional     NaN   
4117  http://www.education.qld.gov.au        NaN          Remote  1116.0   
4295  http://www.education.vic.gov.au        NaN    Major Cities   990.0   
4367    https://education.nsw.gov.au/        NaN    Major Cities     NaN   

      ICSEA Percentile  Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
2736               NaN                     NaN                           NaN   
3626               NaN                     NaN                           NaN   
4117              88.0                     0.0                           0.0   
4295              41.0                    40.0                          39.0   
4367               NaN                     NaN                           NaN   

      Upper Middle SEA Quarter (%)  Top SEA Quarter (%)  Teaching Staff  \
2736                           NaN                  NaN             NaN   
3626                           NaN                  NaN             NaN   
4117                           0.0                100.0             NaN   
4295                          15.0                  7.0             NaN   
4367                           NaN                  NaN             NaN   

      Non-Teaching Staff  Total Enrolments  Girls Enrolments  Boys Enrolments  \
2736                 NaN               NaN               NaN              NaN   
3626                 NaN               NaN               NaN              NaN   
4117                 NaN               NaN               NaN              NaN   
4295                 NaN               NaN               NaN              NaN   
4367                 NaN               NaN               NaN              NaN   

      Indigenous Enrolments (%)  Language Background Other Than English (%)  
2736                        NaN                                         NaN  
3626                        NaN                                         NaN  
4117                        NaN                                         NaN  
4295                        NaN                                         NaN  
4367                        NaN                                         NaN

From the above we can see that there are a number of missing values across schools that do not list a year range.  However we also noticed another interesting fact, the school type will list *Primary, Secondary & Combined* 

Using our knowledge of australian schooling we can infer that primary is pertaining to prep-6 (prep-7 for some years), secondary is equal to 8-12 (7-12 for some years) and lastly combined encompases prep-12. 

Although the above entries do not include the year range they list if they are combined or otherwise.  

Moving foward we will use this metric to gauge the year levels a school caters for. 

So lets get rid of the Year Range column

In [21]:
#assigning the year range column to a variable
new_cols_to_drop = df2_csv[['Year Range']]
new_cols_to_drop

Year Range
0       U, Prep-6
1       U, Prep-6
2       U, Prep-6
3       U, Prep-6
4       U, Prep-6
...           ...
111514        1-7
111515       PP-7
111516     Prep-6
111517     Prep-5
111518         10

[111519 rows x 1 columns]

In [22]:
# Now lets get rid of it and check its gone
# We will keep it in the same dataframe for now
df2_csv = df2_csv.drop(new_cols_to_drop, axis=1)

df2_csv.head()

Calendar Year                     School Name     Suburb State  Postcode  \
0           2008  Corpus Christi Catholic School  Bellerive   TAS      7018   
1           2009  Corpus Christi Catholic School  Bellerive   TAS      7018   
2           2010  Corpus Christi Catholic School  Bellerive   TAS      7018   
3           2011  Corpus Christi Catholic School  Bellerive   TAS      7018   
4           2012  Corpus Christi Catholic School  Bellerive   TAS      7018   

  School Sector School Type           Campus Type  \
0      Catholic     Primary  School Single Entity   
1      Catholic     Primary  School Single Entity   
2      Catholic     Primary  School Single Entity   
3      Catholic     Primary  School Single Entity   
4      Catholic     Primary  School Single Entity   

                            School URL                       Governing Body  \
0  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
1  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
2  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
3  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   
4  http://www.corpuschristi.tas.edu.au  Tasmanian Catholic Education Office   

               Governing Body URL     Geolocation   ICSEA  ICSEA Percentile  \
0  http://www.catholic.tas.edu.au  Inner Regional  1055.0              75.0   
1  http://www.catholic.tas.edu.au  Inner Regional  1055.0              75.0   
2  http://www.catholic.tas.edu.au  Inner Regional  1044.0              70.0   
3  http://www.catholic.tas.edu.au  Inner Regional  1046.0              71.0   
4  http://www.catholic.tas.edu.au  Inner Regional  1041.0              69.0   

   Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
0                    13.0                          19.0   
1                    13.0                          19.0   
2                    28.0                          19.0   
3                    19.0                          30.0   
4                     9.0                          35.0   

   Upper Middle SEA Quarter (%)  Top SEA Quarter (%)  Teaching Staff  \
0                          37.0                 31.0            29.0   
1                          37.0                 31.0            27.0   
2                          30.0                 23.0            26.0   
3                          25.0                 25.0            25.0   
4                          28.0                 28.0            24.0   

   Non-Teaching Staff  Total Enrolments  Girls Enrolments  Boys Enrolments  \
0                12.0             410.0             185.0            225.0   
1                15.0             409.0             184.0            225.0   
2                18.0             392.0             173.0            219.0   
3                16.0             385.0             173.0            212.0   
4                16.0             391.0             185.0            206.0   

   Indigenous Enrolments (%)  Language Background Other Than English (%)  
0                        1.0                                         4.0  
1                        2.0                                         4.0  
2                        2.0                                         4.0  
3                        0.0                                         3.0  
4                        1.0                                         3.0

One step closer, lets look at the cols we have left to clean

In [23]:
# Duplicating code from above to drill down even further
# Storing in new variables just in case 
missing_values1 = df2_csv.isna().sum()

missing_percent1 = ((missing_values1 / df2_csv.shape[0]) * 100)

columns_to_fix1 = list(missing_percent1[(missing_percent1 > 0)].index)
columns_to_fix1

['ICSEA',
 'ICSEA Percentile',
 'Bottom SEA Quarter (%)',
 'Lower Middle SEA Quarter (%)',
 'Upper Middle SEA Quarter (%)',
 'Top SEA Quarter (%)',
 'Teaching Staff',
 'Non-Teaching Staff',
 'Total Enrolments',
 'Girls Enrolments',
 'Boys Enrolments',
 'Indigenous Enrolments (%)',
 'Language Background Other Than English (%)']

Next lets investigate the ICSEA and ICSEA Percentile columns.  These will be used to gauge learning outcomes and missing values doesnt give an accurate indication of the performance of the school when using the ICSEA metric. 

In [24]:
icsea_missing = df2_csv[df2_csv['ICSEA'].isna()]
icsea_missing.head(20)

Calendar Year                 School Name               Suburb State  \
742            2018        Ouse District School                 Ouse   TAS   
2430           2014  MacKillop Catholic College           Mornington   TAS   
2431           2015  MacKillop Catholic College           Mornington   TAS   
2460           2008        Giant Steps Tasmania            Deloraine   TAS   
2461           2009        Giant Steps Tasmania            Deloraine   TAS   
2462           2010        Giant Steps Tasmania            Deloraine   TAS   
2463           2011        Giant Steps Tasmania            Deloraine   TAS   
2464           2012        Giant Steps Tasmania            Deloraine   TAS   
2465           2013        Giant Steps Tasmania            Deloraine   TAS   
2466           2014        Giant Steps Tasmania            Deloraine   TAS   
2467           2015        Giant Steps Tasmania            Deloraine   TAS   
2469           2017        Giant Steps Tasmania            Deloraine   TAS   
2658           2014   Cape Barren Island School   Cape Barren Island   TAS   
2659           2015   Cape Barren Island School   Cape Barren Island   TAS   
2660           2016   Cape Barren Island School   Cape Barren Island   TAS   
2663           2019   Cape Barren Island School   Cape Barren Island   TAS   
2694           2014  Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2695           2015  Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2696           2016  Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2697           2017  Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   

      Postcode School Sector School Type           Campus Type  \
742       7140    Government     Primary  School Single Entity   
2430      7018      Catholic   Secondary  School Single Entity   
2431      7018      Catholic   Secondary  School Single Entity   
2460      7304   Independent     Special  School Single Entity   
2461      7304   Independent     Special  School Single Entity   
2462      7304   Independent     Special  School Single Entity   
2463      7304   Independent     Special  School Single Entity   
2464      7304   Independent     Special  School Single Entity   
2465      7304   Independent     Special  School Single Entity   
2466      7304   Independent     Special  School Single Entity   
2467      7304   Independent     Special  School Single Entity   
2469      7304   Independent     Special  School Single Entity   
2658      7257    Government    Combined  School Single Entity   
2659      7257    Government    Combined  School Single Entity   
2660      7257    Government    Combined  School Single Entity   
2663      7257    Government    Combined  School Single Entity   
2694       872    Government    Combined    School Head Campus   
2695       872    Government    Combined    School Head Campus   
2696       872    Government    Combined    School Head Campus   
2697       872    Government    Combined    School Head Campus   

                                             School URL  \
742   http://education.tas.edu.au/ousedistrict/SiteP...   
2430                    http://www.mackillop.tas.edu.au   
2431                    http://www.mackillop.tas.edu.au   
2460                  http://www.giantsteps.tas.edu.au/   
2461                  http://www.giantsteps.tas.edu.au/   
2462                  http://www.giantsteps.tas.edu.au/   
2463                  http://www.giantsteps.tas.edu.au/   
2464                  http://www.giantsteps.tas.edu.au/   
2465                  http://www.giantsteps.tas.edu.au/   
2466                  http://www.giantsteps.tas.edu.au/   
2467                  http://www.giantsteps.tas.edu.au/   
2469                  http://www.giantsteps.tas.edu.au/   
2658  http://www.education.tas.gov.au/dept/contact/s...   
2659  http://www.education.tas.gov.au/dept/contact/s...   
2660  http://www.education.tas.gov.au/dept/contact/s...   
2663  http://www.education.tas.gov.a

Using the above findings we can indeed see that some of the schools do not report on their ICSEA score, there may be a multitude of reasons for this however in our use case we need to quantatively measure the educational outcomes and this is the closest thing we have.  These rows will have to go so lets create a variable to get rid of them

In [25]:
icsea_missing_values = df2_csv[df2_csv['ICSEA'].isna()]
icsea_missing_values

Calendar Year                                    School Name  \
742              2018                           Ouse District School   
2430             2014                     MacKillop Catholic College   
2431             2015                     MacKillop Catholic College   
2460             2008                           Giant Steps Tasmania   
2461             2009                           Giant Steps Tasmania   
...               ...                                            ...   
111482           2019                       Indie School - Glenorchy   
111500           2019                                SEDA College SA   
111512           2019                        Ontrack College Emerald   
111514           2019                            Nawarddeken Academy   
111518           2019  Australian Industry Trade College - Toowoomba   

                  Suburb State  Postcode School Sector School Type  \
742                 Ouse   TAS      7140    Government     Primary   
2430          Mornington   TAS      7018      Catholic   Secondary   
2431          Mornington   TAS      7018      Catholic   Secondary   
2460           Deloraine   TAS      7304   Independent     Special   
2461           Deloraine   TAS      7304   Independent     Special   
...                  ...   ...       ...           ...         ...   
111482         Glenorchy   TAS      7010   Independent   Secondary   
111500          Glandore    SA      5037   Independent   Secondary   
111512           Emerald   QLD      4720   Independent     Special   
111514  West Arnhem Land    NT       822   Independent    Combined   
111518    Toowoomba City   QLD      4350   Independent   Secondary   

                 Campus Type  \
742     School Single Entity   
2430    School Single Entity   
2431    School Single Entity   
2460    School Single Entity   
2461    School Single Entity   
...                      ...   
111482  School Single Entity   
111500  School Single Entity   
111512  School Single Entity   
111514  School Single Entity   
111518  School Single Entity   

                                               School URL  \
742     http://education.tas.edu.au/ousedistrict/SiteP...   
2430                      http://www.mackillop.tas.edu.au   
2431                      http://www.mackillop.tas.edu.au   
2460                    http://www.giantsteps.tas.edu.au/   
2461                    http://www.giantsteps.tas.edu.au/   
...                                                   ...   
111482                                            unknown   
111500                          http://www.seda.sa.edu.au   
111512                        https://www.pcycnsw.org.au/   
111514                 https://www.nawarddekenacademy.com   
111518                                            unknown   

                                           Governing Body  \
742                      Department of Education Tasmania   
2430                  Tasmanian Catholic Education Office   
2431                  Tasmanian Catholic Education Office   
2460                         Independent Schools Tasmania   
2461                         Independent Schools Tasmania   
...                                                   ...   
111482                       Independent Schools Tasmania   
111500  Association of Independent Schools of South Au...   
111512                     Independent Schools Queensland   
111514  Association of Independent Schools of the Nort...   
111518                     Independent Schools Queensland   

                              Governing Body URL     Geolocation  ICSEA  \
742              http://www.education.tas.gov.au  Outer Regional    NaN   
2430              http://www.catholic.tas.edu.au  Inner Regional    NaN   
2431              http://www.catholic.tas.edu.au  Inner Regional    NaN   
2460    http://www.independentschools.tas.edu.au  Outer Regional    NaN   
2461    http://www.independentschools.tas.edu.au  Outer Regional    NaN   
.

In [26]:
rows_to_drop = list(icsea_missing_values.index)

rows_to_drop[:5]

[742, 2430, 2431, 2460, 2461]

It appears they are ready to drop, so lets get rid of them

In [27]:
# Saving in a new DF as the old one is getting well used
# This also creates a safety checkpoint of sorts

df3 = df2_csv.drop(rows_to_drop, axis=0)

For fun lets compare where we started to where we are now

In [28]:
progress = len(df1_csv) - len(df3)

print(f'We have reduced our dataset by {progress} entries! woohoo!')

We have reduced our dataset by 6060 entries! woohoo!


Quick status report!

In [29]:
# Duplicating code from above to drill down even further
# Storing in new variables just in case 
missing_values2 = df3.isna().sum()

missing_percent2 = ((missing_values2 / df3.shape[0]) * 100)

columns_to_fix2 = list(missing_percent2[(missing_percent2 > 0)].index)
columns_to_fix2

['ICSEA Percentile',
 'Bottom SEA Quarter (%)',
 'Lower Middle SEA Quarter (%)',
 'Upper Middle SEA Quarter (%)',
 'Top SEA Quarter (%)',
 'Teaching Staff',
 'Non-Teaching Staff',
 'Total Enrolments',
 'Girls Enrolments',
 'Boys Enrolments',
 'Indigenous Enrolments (%)',
 'Language Background Other Than English (%)']

One thing noticed was that Teaching and non teaching staff also have values missing.  Lets check it out. 

In [30]:
teachers_missing = df3[df3['Teaching Staff'].isna()]
len(teachers_missing)

99

Using the above we can see that there are 99 entries with missing teacher data. This is insignificant to the overall picture so we can safely discard these. Let us check if the non teaching staff paints the same picture or is slightly different

In [31]:
non_teachers_missing = df3[df3['Non-Teaching Staff'].isna()]
len(non_teachers_missing)

118

It is different meaning there may be some schools who have teachers but no non teaching staff.  This would be more common in the very remote regions of Australia with small school sizes etc.  We can safely remove these rows from the dataframe. 

In [32]:
t_list = list(teachers_missing.index)
nt_list = list(non_teachers_missing.index)

more_rows_to_drop = list(set(t_list + nt_list))
len(more_rows_to_drop)

118

In [33]:
# Getting rid of the rows
df3 = df3.drop(more_rows_to_drop, axis=0)

df3.shape

(105341, 25)

We are getting there, but we still have a long way to go.  Also another thing to note while cleaning we noticed that there are still missing values in the ICSEA percentile, lets investigate these further

In [34]:
icsea_per_missing_values = df3[df3['ICSEA Percentile'].isna()]
icsea_per_missing_values

Calendar Year                               School Name        Suburb  \
2471             2019                      Giant Steps Tasmania     Deloraine   
2805             2019                    Sir David Brand School     Coolbinia   
2841             2019                  Thebarton Senior College  Torrensville   
2855             2019               O'Connor Cooperative School      O'Connor   
2867             2019                          Erindale College     Wanniassa   
...               ...                                       ...           ...   
111503           2019                  Yawarra Community School         Dubbo   
111507           2019   Santa Sophia Catholic College (Primary)    Kellyville   
111508           2019                   Maxwell Creative School      Cockatoo   
111510           2019                   Enkindle Village School       Douglas   
111513           2019  Carbrook Animal Assisted Learning Centre      Carbrook   

       State  Postcode School Sector School Type           Campus Type  \
2471     TAS      7304   Independent     Special  School Single Entity   
2805      WA      6050    Government     Special  School Single Entity   
2841      SA      5031    Government   Secondary  School Single Entity   
2855     ACT      2602    Government     Primary  School Single Entity   
2867     ACT      2903    Government   Secondary  School Single Entity   
...      ...       ...           ...         ...                   ...   
111503   NSW      2830    Government     Special  School Single Entity   
111507   NSW      2155      Catholic     Primary  School Single Entity   
111508   VIC      3781   Independent     Primary  School Single Entity   
111510   QLD      4814   Independent     Primary  School Single Entity   
111513   QLD      4130   Independent   Secondary  School Single Entity   

                                               School URL  \
2471                    http://www.giantsteps.tas.edu.au/   
2805    http://www.det.wa.edu.au/schoolsonline/overvie...   
2841                             http://www.tsc.sa.edu.au   
2855                          http://www.coop.act.edu.au/   
2867                     http://www.erindalec.act.edu.au/   
...                                                   ...   
111503              https://yawarra-c.schools.nsw.gov.au/   
111507                                            unknown   
111508                                            unknown   
111510                               http://www.yahoo.com   
111513                                            unknown   

                                   Governing Body  \
2471                 Independent Schools Tasmania   
2805    Department of Education Western Australia   
2841     Department for Education South Australia   
2855                    ACT Education Directorate   
2867                    ACT Education Directorate   
...                                           ...   
111503                Department of Education NSW   
111507                       Catholic Schools NSW   
111508               Independent Schools Victoria   
111510             Independent Schools Queensland   
111513             Independent Schools Queensland   

                              Governing Body URL     Geolocation   ICSEA  \
2471    http://www.independentschools.tas.edu.au  Outer Regional   949.0   
2805                    http://www.det.wa.edu.au    Major Cities  1001.0   
2841            https://www.education.sa.gov.au/    Major Cities   978.0   
2855                   http://www.det.act.gov.au    Major Cities  1133.0   
2867                   http://www.det.act.gov.au    Major Cities  1022.0   
...                                          ...             ...     ...   
111503             https://education.nsw.gov.au/  Inner Regional   908.0   
111507          http://www.csnsw.catholic.edu.au    Major Cities  1114.0   
111508                 https://www.is.vic.edu.au    Major Cities  1035.0   
111510          

Looking at the above we need to evaluate if the percentile is as important as the score.  Drilling down the information we can see that the overall percentile paints a broarder picture than the specific bottom, lower, middle and upper percentile figures found in the SEA columns.  We will be better off using this to gauge the learning outcomes vs an overall percentile figure as we need more accuracy when comparing schools. 

Another thing to note is that while the governing body information is a metric we may be interested in, the web URL associated with it wont be.  So Let's remove both of them now. 

In [35]:
#assigning the columns to a variable and getting rid of them
new_col_to_drop = df3[['ICSEA Percentile', 'Governing Body URL']]

df3= df3.drop(new_col_to_drop, axis=1)

df3.isna().sum() # Checking how we are doing

Calendar Year                                    0
School Name                                      0
Suburb                                           0
State                                            0
Postcode                                         0
School Sector                                    0
School Type                                      0
Campus Type                                      0
School URL                                       0
Governing Body                                   0
Geolocation                                      0
ICSEA                                            0
Bottom SEA Quarter (%)                        4187
Lower Middle SEA Quarter (%)                  4187
Upper Middle SEA Quarter (%)                  4187
Top SEA Quarter (%)                           4187
Teaching Staff                                   0
Non-Teaching Staff                               0
Total Enrolments                                 4
Girls Enrolments               

Looking at the above we can see a few interesting things:
- All of the SEA quarter % values are the same, we can infer that all 4187 entries are missing all SEA quarter values and so we can remove them.
- Total enrolments, Girl and Boy enrolments are tiny and we can look at these individually to check if we need to keep these entries
- Indigenous Enrolments and Language background other than english are still left and we will deal with these accordingly. 

let us check out the enrolments now

In [36]:
# Checking the girls enrolments as this encompases 5 entries rather than 4 entries
enrol_missing_values = df3[df3['Girls Enrolments'].isna()]
enrol_missing_values

Calendar Year                            School Name          Suburb  \
5037            2018                  Royal Far West School           Manly   
5397            2018  Queensland Children's Hospital School  South Brisbane   
5613            2018       Sydney Childrens Hospital School        Randwick   
39473           2011           St Andrew's Cathedral School          Sydney   
43857           2018                 Navarre Primary School         Navarre   

      State  Postcode School Sector School Type           Campus Type  \
5037    NSW      2095    Government     Special  School Single Entity   
5397    QLD      4101    Government     Special    School Head Campus   
5613    NSW      2031    Government     Special  School Single Entity   
39473   NSW      2000   Independent    Combined    School Head Campus   
43857   VIC      3384    Government     Primary  School Single Entity   

                                       School URL  \
5037     https://royalfarwe-s.schools.nsw.gov.au/   
5397               http://www.qchschool.eq.edu.au   
5613      https://sydchnhos-s.schools.nsw.gov.au/   
39473                  http://www.sacs.nsw.edu.au   
43857  http://www.navarreprimaryschool.weebly.com   

                                          Governing Body     Geolocation  \
5037                         Department of Education NSW    Major Cities   
5397                  Department of Education Queensland    Major Cities   
5613                         Department of Education NSW    Major Cities   
39473  Association of Independent Schools of New Sout...    Major Cities   
43857      Department of Education and Training Victoria  Outer Regional   

        ICSEA  Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
5037    960.0                    10.0                          30.0   
5397    908.0                    43.0                          23.0   
5613   1038.0                    10.0                          40.0   
39473  1180.0                     NaN                           NaN   
43857   987.0                    33.0                          33.0   

       Upper Middle SEA Quarter (%)  Top SEA Quarter (%)  Teaching Staff  \
5037                           30.0                 30.0             7.0   
5397                           23.0                 10.0            39.0   
5613                           40.0                 10.0             8.0   
39473                           NaN                  NaN           113.0   
43857                          30.0                  3.0             2.0   

       Non-Teaching Staff  Total Enrolments  Girls Enrolments  \
5037                  9.0               NaN               NaN   
5397                 22.0               NaN               NaN   
5613                  5.0               NaN               NaN   
39473                73.0            1087.0               NaN   
43857                 1.0               NaN               NaN   

       Boys Enrolments  Indigenous Enrolments (%)  \
5037               NaN                        NaN   
5397               NaN                        NaN   
5613               NaN                        NaN   
39473              NaN                        2.0   
43857              NaN                        NaN   

       Language Background Other Than English (%)  
5037                                          NaN  
5397                                          NaN  
5613                                          NaN  
39473                                        23.0  
43857                                         NaN

We can see that the 5 missing boy and girl enrolment values are from the same school, however we can also see that there is one entry which still has a total enrolments value available.  

Moving forward what we will do is remove the schools with no enrolment data, and then sort the remaining entry out by comparing it to other entries from the same school. 

In [37]:
# Gathering and dropping the rows with no enrolment information at all
total_enrol_missing_values = df3[df3['Total Enrolments'].isna()]

# Collating them into a list
rows_to_drop = list(total_enrol_missing_values.index)

# Checking the list
rows_to_drop[:5]

[5037, 5397, 5613, 43857]

In [38]:
# Drop em!
df3 = df3.drop(rows_to_drop, axis=0)

# How are we doing?
df3.isna().sum()

Calendar Year                                    0
School Name                                      0
Suburb                                           0
State                                            0
Postcode                                         0
School Sector                                    0
School Type                                      0
Campus Type                                      0
School URL                                       0
Governing Body                                   0
Geolocation                                      0
ICSEA                                            0
Bottom SEA Quarter (%)                        4187
Lower Middle SEA Quarter (%)                  4187
Upper Middle SEA Quarter (%)                  4187
Top SEA Quarter (%)                           4187
Teaching Staff                                   0
Non-Teaching Staff                               0
Total Enrolments                                 0
Girls Enrolments               

The next logical step would be to check out the girls enrolments and boys enrolments entry to see what the story is, however from the above cell we can see that this particular school doesnt have any SEA Quarter values at all. So we will deal with them first to see how that affects the remaining amount to be cleaned. 

In [39]:
# Checking the bottom SEA Quarter value
sea_missing_values = df3[df3['Bottom SEA Quarter (%)'].isna()]
sea_missing_values

Calendar Year                        School Name               Suburb  \
63               2011                 John Calvin School           Launceston   
231              2011        St Joseph's Catholic School             Rosebery   
351              2011                 The Cottage School            Bellerive   
361              2009        Launceston Christian School            Riverside   
435              2011  The Launceston Preparatory School           Launceston   
...               ...                                ...                  ...   
109728           2011   Springfield Central State School  Springfield Central   
109755           2011        Coomera Rivers State School              Coomera   
109764           2011         Makybe Rise Primary School             Baldivis   
110716           2011               Radiant Life College       East Innisfail   
110717           2012               Radiant Life College       East Innisfail   

       State  Postcode School Sector School Type           Campus Type  \
63       TAS      7250   Independent    Combined  School Single Entity   
231      TAS      7470      Catholic     Primary  School Single Entity   
351      TAS      7018   Independent     Primary  School Single Entity   
361      TAS      7250   Independent    Combined  School Single Entity   
435      TAS      7250   Independent     Primary  School Single Entity   
...      ...       ...           ...         ...                   ...   
109728   QLD      4300    Government     Primary  School Single Entity   
109755   QLD      4209    Government     Primary  School Single Entity   
109764    WA      6171    Government     Primary  School Single Entity   
110716   QLD      4860   Independent    Combined  School Single Entity   
110717   QLD      4860   Independent    Combined  School Single Entity   

                                               School URL  \
63                              http://www.jcs.tas.edu.au   
231                      http://www.stjosephsr.tas.edu.au   
351                   http://www.cottageschool.tas.edu.au   
361                             http://www.lcs.tas.edu.au   
435                             http://www.lps.tas.edu.au   
...                                                   ...   
109728          http://www.springfieldcentralss.eq.edu.au   
109755               http://www.coomerariversss.eq.edu.au   
109764  http://www.det.wa.edu.au/schoolsonline/overvie...   
110716                      http://radiantlife.qld.edu.au   
110717                      http://radiantlife.qld.edu.au   

                                   Governing Body     Geolocation   ICSEA  \
63                   Independent Schools Tasmania  Inner Regional  1027.0   
231           Tasmanian Catholic Education Office          Remote   959.0   
351                  Independent Schools Tasmania  Inner Regional  1054.0   
361                  Independent Schools Tasmania  Inner Regional  1042.0   
435                  Independent Schools Tasmania  Inner Regional  1123.0   
...                                           ...             ...     ...   
109728         Department of Education Queensland    Major Cities  1029.0   
109755         Department of Education Queensland    Major Cities  1039.0   
109764  Department of Education Western Australia    Major Cities   999.0   
110716             Independent Schools Queensland  Outer Regional   612.0   
110717             Independent Schools Queensland  Outer Regional   618.0   

        Bottom SEA Quarter (%)  Lower Middle SEA Quarter (%)  \
63                         NaN                           NaN   
231                        NaN                           NaN   
351                        NaN                           NaN   
361                        NaN                           NaN   
435                        NaN                           NaN   
...                        ...                           ...   
109728                     NaN          

Looking at the value we can see that while the schools are missing their drilled down percentiles, they do still have their overall ICSEA score for the year.  We are at a crossroads, do we:
- Remove all bottom lower middle and top percentile columns?
- leave them and fill with null values due to there being no other data to formulate a mean value for these entries

Unfortunately populating them with 0's will skew the data in a way that is not favorable to the end user, some of these schools have an ICSEA score above 1000 (The national average) and therefore according to the ICSEA measure are schools that perform well. 

To answer the questions we have, the overall ICSEA value is far more important as it encompases an overall score for the school, a definitive measure that we can utilise to compare schools against each other. 

Lets get rid of the 4 columns we dont need

In [40]:
#assigning the columns to a variable and getting rid of them
sea_cols_to_drop = df3[['Bottom SEA Quarter (%)', 'Lower Middle SEA Quarter (%)', 'Upper Middle SEA Quarter (%)', 'Top SEA Quarter (%)']]

# Assigning it to a new data frame
df4= df3.drop(sea_cols_to_drop, axis=1)

df4.isna().sum() # Check how we are doing

Calendar Year                                    0
School Name                                      0
Suburb                                           0
State                                            0
Postcode                                         0
School Sector                                    0
School Type                                      0
Campus Type                                      0
School URL                                       0
Governing Body                                   0
Geolocation                                      0
ICSEA                                            0
Teaching Staff                                   0
Non-Teaching Staff                               0
Total Enrolments                                 0
Girls Enrolments                                 1
Boys Enrolments                                  1
Indigenous Enrolments (%)                     3067
Language Background Other Than English (%)    9307
dtype: int64

We are nearly there!  let us investigate the boy and girl enrolment by checking if they are the same value


In [41]:
girls_enrol_missing = df4[df4['Girls Enrolments'].isna()]
girls_enrol_missing.head()

Calendar Year                   School Name  Suburb State  Postcode  \
39473           2011  St Andrew's Cathedral School  Sydney   NSW      2000   

      School Sector School Type         Campus Type  \
39473   Independent    Combined  School Head Campus   

                       School URL  \
39473  http://www.sacs.nsw.edu.au   

                                          Governing Body   Geolocation  \
39473  Association of Independent Schools of New Sout...  Major Cities   

        ICSEA  Teaching Staff  Non-Teaching Staff  Total Enrolments  \
39473  1180.0           113.0                73.0            1087.0   

       Girls Enrolments  Boys Enrolments  Indigenous Enrolments (%)  \
39473               NaN              NaN                        2.0   

       Language Background Other Than English (%)  
39473                                        23.0

In [42]:
# Let us check if the school has any other entries with enrolment data available
df4[df4['School Name'] == "St Andrew's Cathedral School"].head()

Calendar Year                   School Name  Suburb State  Postcode  \
39470           2008  St Andrew's Cathedral School  Sydney   NSW      2000   
39471           2009  St Andrew's Cathedral School  Sydney   NSW      2000   
39472           2010  St Andrew's Cathedral School  Sydney   NSW      2000   
39473           2011  St Andrew's Cathedral School  Sydney   NSW      2000   
39475           2013  St Andrew's Cathedral School  Sydney   NSW      2000   

      School Sector School Type         Campus Type  \
39470   Independent    Combined  School Head Campus   
39471   Independent    Combined  School Head Campus   
39472   Independent    Combined  School Head Campus   
39473   Independent    Combined  School Head Campus   
39475   Independent    Combined  School Head Campus   

                       School URL  \
39470  http://www.sacs.nsw.edu.au   
39471  http://www.sacs.nsw.edu.au   
39472  http://www.sacs.nsw.edu.au   
39473  http://www.sacs.nsw.edu.au   
39475  http://www.sacs.nsw.edu.au   

                                          Governing Body   Geolocation  \
39470  Association of Independent Schools of New Sout...  Major Cities   
39471  Association of Independent Schools of New Sout...  Major Cities   
39472  Association of Independent Schools of New Sout...  Major Cities   
39473  Association of Independent Schools of New Sout...  Major Cities   
39475  Association of Independent Schools of New Sout...  Major Cities   

        ICSEA  Teaching Staff  Non-Teaching Staff  Total Enrolments  \
39470  1157.0           128.0                64.0            1245.0   
39471  1157.0           126.0                83.0            1233.0   
39472  1173.0           122.0                78.0            1180.0   
39473  1180.0           113.0                73.0            1087.0   
39475  1169.0           110.0                73.0            1043.0   

       Girls Enrolments  Boys Enrolments  Indigenous Enrolments (%)  \
39470             322.0            923.0                        2.0   
39471             362.0            871.0                        3.0   
39472             372.0            808.0                        3.0   
39473               NaN              NaN                        2.0   
39475             345.0            698.0                        2.0   

       Language Background Other Than English (%)  
39470                                        13.0  
39471                                        17.0  
39472                                        22.0  
39473                                        23.0  
39475                                        23.0

As we predicted, this school does have boy and girl enrolment data for previous and future years, lets leverage that to fill in the missing values. The simplest way to do this will be to get the mean of the year prior, and the year after the missing values.  Before we enter numerical values into the columns, lets first check the data type of our remaining columns

In [43]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 105337 entries, 0 to 111517
Data columns (total 19 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Calendar Year                               105337 non-null  int64  
 1   School Name                                 105337 non-null  object 
 2   Suburb                                      105337 non-null  object 
 3   State                                       105337 non-null  object 
 4   Postcode                                    105337 non-null  int64  
 5   School Sector                               105337 non-null  object 
 6   School Type                                 105337 non-null  object 
 7   Campus Type                                 105337 non-null  object 
 8   School URL                                  105337 non-null  object 
 9   Governing Body                              105337 non-null  object 
 

Excellent, they are all the correct data types for what we want. Lets move on

In [44]:
# get the mean of the columns so we can use that data for our NaN values

# First we will reindex the dataframe into a new dataframe so that we can backtrack if needed
df5 = df4.reset_index(drop=True)

# Let's verify if the index has indeed changed
df5[df5['School Name'] == "St Andrew's Cathedral School"].head()

Calendar Year                   School Name  Suburb State  Postcode  \
36755           2008  St Andrew's Cathedral School  Sydney   NSW      2000   
36756           2009  St Andrew's Cathedral School  Sydney   NSW      2000   
36757           2010  St Andrew's Cathedral School  Sydney   NSW      2000   
36758           2011  St Andrew's Cathedral School  Sydney   NSW      2000   
36759           2013  St Andrew's Cathedral School  Sydney   NSW      2000   

      School Sector School Type         Campus Type  \
36755   Independent    Combined  School Head Campus   
36756   Independent    Combined  School Head Campus   
36757   Independent    Combined  School Head Campus   
36758   Independent    Combined  School Head Campus   
36759   Independent    Combined  School Head Campus   

                       School URL  \
36755  http://www.sacs.nsw.edu.au   
36756  http://www.sacs.nsw.edu.au   
36757  http://www.sacs.nsw.edu.au   
36758  http://www.sacs.nsw.edu.au   
36759  http://www.sacs.nsw.edu.au   

                                          Governing Body   Geolocation  \
36755  Association of Independent Schools of New Sout...  Major Cities   
36756  Association of Independent Schools of New Sout...  Major Cities   
36757  Association of Independent Schools of New Sout...  Major Cities   
36758  Association of Independent Schools of New Sout...  Major Cities   
36759  Association of Independent Schools of New Sout...  Major Cities   

        ICSEA  Teaching Staff  Non-Teaching Staff  Total Enrolments  \
36755  1157.0           128.0                64.0            1245.0   
36756  1157.0           126.0                83.0            1233.0   
36757  1173.0           122.0                78.0            1180.0   
36758  1180.0           113.0                73.0            1087.0   
36759  1169.0           110.0                73.0            1043.0   

       Girls Enrolments  Boys Enrolments  Indigenous Enrolments (%)  \
36755             322.0            923.0                        2.0   
36756             362.0            871.0                        3.0   
36757             372.0            808.0                        3.0   
36758               NaN              NaN                        2.0   
36759             345.0            698.0                        2.0   

       Language Background Other Than English (%)  
36755                                        13.0  
36756                                        17.0  
36757                                        22.0  
36758                                        23.0  
36759                                        23.0

In [45]:
# Calculating the mean of the columns
boys_mean = df5[['Boys Enrolments']].iloc[[36757, 36759]].mean(axis=0)
girls_mean = df5[['Girls Enrolments']].iloc[[36757, 36759]].mean(axis=0)

print(f'The mean of the Boys enrolments is {boys_mean}')
print(f'The mean of the Girls enrolments is {girls_mean}')

The mean of the Boys enrolments is Boys Enrolments    753.0
dtype: float64
The mean of the Girls enrolments is Girls Enrolments    358.5
dtype: float64


In [46]:
boys_mean.dtype

dtype('float64')

In [47]:
# Trying to replace the values with the means
df5.at[36758, 'Boys Enrolments'] = boys_mean
df5.at[36758, 'Girls Enrolments'] = girls_mean
# Checking to see if it worked
df5[df5['School Name'] == "St Andrew's Cathedral School"].head()

Calendar Year                   School Name  Suburb State  Postcode  \
36755           2008  St Andrew's Cathedral School  Sydney   NSW      2000   
36756           2009  St Andrew's Cathedral School  Sydney   NSW      2000   
36757           2010  St Andrew's Cathedral School  Sydney   NSW      2000   
36758           2011  St Andrew's Cathedral School  Sydney   NSW      2000   
36759           2013  St Andrew's Cathedral School  Sydney   NSW      2000   

      School Sector School Type         Campus Type  \
36755   Independent    Combined  School Head Campus   
36756   Independent    Combined  School Head Campus   
36757   Independent    Combined  School Head Campus   
36758   Independent    Combined  School Head Campus   
36759   Independent    Combined  School Head Campus   

                       School URL  \
36755  http://www.sacs.nsw.edu.au   
36756  http://www.sacs.nsw.edu.au   
36757  http://www.sacs.nsw.edu.au   
36758  http://www.sacs.nsw.edu.au   
36759  http://www.sacs.nsw.edu.au   

                                          Governing Body   Geolocation  \
36755  Association of Independent Schools of New Sout...  Major Cities   
36756  Association of Independent Schools of New Sout...  Major Cities   
36757  Association of Independent Schools of New Sout...  Major Cities   
36758  Association of Independent Schools of New Sout...  Major Cities   
36759  Association of Independent Schools of New Sout...  Major Cities   

        ICSEA  Teaching Staff  Non-Teaching Staff  Total Enrolments  \
36755  1157.0           128.0                64.0            1245.0   
36756  1157.0           126.0                83.0            1233.0   
36757  1173.0           122.0                78.0            1180.0   
36758  1180.0           113.0                73.0            1087.0   
36759  1169.0           110.0                73.0            1043.0   

       Girls Enrolments  Boys Enrolments  Indigenous Enrolments (%)  \
36755             322.0            923.0                        2.0   
36756             362.0            871.0                        3.0   
36757             372.0            808.0                        3.0   
36758             358.5            753.0                        2.0   
36759             345.0            698.0                        2.0   

       Language Background Other Than English (%)  
36755                                        13.0  
36756                                        17.0  
36757                                        22.0  
36758                                        23.0  
36759                                        23.0

This worked, sort of, but it is not ideal.  The two values of 358.5 and 753 do not add up to the confirmed total of 1087 enrolments.  We are 24.5 students above this.  Moving forward we can split them evenly and subtract 12.5 from the girls enrolments, and 12 from the boys enrolments to get a more accurate figure. 

In [48]:
# modifying mean values
df5.at[36758, 'Boys Enrolments'] = boys_mean - 12
df5.at[36758, 'Girls Enrolments'] = girls_mean - 12.5
# Checking to see if it worked
df5[df5['School Name'] == "St Andrew's Cathedral School"].head()

Calendar Year                   School Name  Suburb State  Postcode  \
36755           2008  St Andrew's Cathedral School  Sydney   NSW      2000   
36756           2009  St Andrew's Cathedral School  Sydney   NSW      2000   
36757           2010  St Andrew's Cathedral School  Sydney   NSW      2000   
36758           2011  St Andrew's Cathedral School  Sydney   NSW      2000   
36759           2013  St Andrew's Cathedral School  Sydney   NSW      2000   

      School Sector School Type         Campus Type  \
36755   Independent    Combined  School Head Campus   
36756   Independent    Combined  School Head Campus   
36757   Independent    Combined  School Head Campus   
36758   Independent    Combined  School Head Campus   
36759   Independent    Combined  School Head Campus   

                       School URL  \
36755  http://www.sacs.nsw.edu.au   
36756  http://www.sacs.nsw.edu.au   
36757  http://www.sacs.nsw.edu.au   
36758  http://www.sacs.nsw.edu.au   
36759  http://www.sacs.nsw.edu.au   

                                          Governing Body   Geolocation  \
36755  Association of Independent Schools of New Sout...  Major Cities   
36756  Association of Independent Schools of New Sout...  Major Cities   
36757  Association of Independent Schools of New Sout...  Major Cities   
36758  Association of Independent Schools of New Sout...  Major Cities   
36759  Association of Independent Schools of New Sout...  Major Cities   

        ICSEA  Teaching Staff  Non-Teaching Staff  Total Enrolments  \
36755  1157.0           128.0                64.0            1245.0   
36756  1157.0           126.0                83.0            1233.0   
36757  1173.0           122.0                78.0            1180.0   
36758  1180.0           113.0                73.0            1087.0   
36759  1169.0           110.0                73.0            1043.0   

       Girls Enrolments  Boys Enrolments  Indigenous Enrolments (%)  \
36755             322.0            923.0                        2.0   
36756             362.0            871.0                        3.0   
36757             372.0            808.0                        3.0   
36758             346.0            741.0                        2.0   
36759             345.0            698.0                        2.0   

       Language Background Other Than English (%)  
36755                                        13.0  
36756                                        17.0  
36757                                        22.0  
36758                                        23.0  
36759                                        23.0

We must be close now...

In [49]:
df5.isna().sum()

Calendar Year                                    0
School Name                                      0
Suburb                                           0
State                                            0
Postcode                                         0
School Sector                                    0
School Type                                      0
Campus Type                                      0
School URL                                       0
Governing Body                                   0
Geolocation                                      0
ICSEA                                            0
Teaching Staff                                   0
Non-Teaching Staff                               0
Total Enrolments                                 0
Girls Enrolments                                 0
Boys Enrolments                                  0
Indigenous Enrolments (%)                     3067
Language Background Other Than English (%)    9307
dtype: int64

Using similar methods as above, lets have a look and see if the indigenous enrolments are not entered or if they should be zero. 

In [50]:
indigenous_missing = df5[df5['Indigenous Enrolments (%)'].isna()]
indigenous_missing.head(10)

Calendar Year         School Name      Suburb State  Postcode  \
64            2012  John Calvin School  Launceston   TAS      7250   
65            2013  John Calvin School  Launceston   TAS      7250   
66            2014  John Calvin School  Launceston   TAS      7250   
67            2015  John Calvin School  Launceston   TAS      7250   
68            2016  John Calvin School  Launceston   TAS      7250   
352           2012  The Cottage School   Bellerive   TAS      7018   
353           2013  The Cottage School   Bellerive   TAS      7018   
354           2014  The Cottage School   Bellerive   TAS      7018   
355           2015  The Cottage School   Bellerive   TAS      7018   
356           2016  The Cottage School   Bellerive   TAS      7018   

    School Sector School Type           Campus Type  \
64    Independent    Combined  School Single Entity   
65    Independent    Combined  School Single Entity   
66    Independent    Combined  School Single Entity   
67    Independent    Combined  School Single Entity   
68    Independent    Combined  School Single Entity   
352   Independent     Primary  School Single Entity   
353   Independent     Primary  School Single Entity   
354   Independent     Primary  School Single Entity   
355   Independent     Primary  School Single Entity   
356   Independent     Primary  School Single Entity   

                              School URL                Governing Body  \
64             http://www.jcs.tas.edu.au  Independent Schools Tasmania   
65             http://www.jcs.tas.edu.au  Independent Schools Tasmania   
66             http://www.jcs.tas.edu.au  Independent Schools Tasmania   
67             http://www.jcs.tas.edu.au  Independent Schools Tasmania   
68             http://www.jcs.tas.edu.au  Independent Schools Tasmania   
352  http://www.cottageschool.tas.edu.au  Independent Schools Tasmania   
353  http://www.cottageschool.tas.edu.au  Independent Schools Tasmania   
354  http://www.cottageschool.tas.edu.au  Independent Schools Tasmania   
355  http://www.cottageschool.tas.edu.au  Independent Schools Tasmania   
356  http://www.cottageschool.tas.edu.au  Independent Schools Tasmania   

        Geolocation   ICSEA  Teaching Staff  Non-Teaching Staff  \
64   Inner Regional  1032.0            10.0                 4.0   
65   Inner Regional  1026.0            10.0                 4.0   
66   Inner Regional  1032.0            16.0                 4.0   
67   Inner Regional  1034.0            11.0                 7.0   
68   Inner Regional  1037.0            12.0                 7.0   
352  Inner Regional  1060.0             9.0                 6.0   
353  Inner Regional  1113.0             9.0                 8.0   
354  Inner Regional  1086.0            10.0                 7.0   
355  Inner Regional  1096.0            10.0                 7.0   
356  Inner Regional  1092.0             9.0                 6.0   

     Total Enrolments  Girls Enrolments  Boys Enrolments  \
64               69.0              34.0             35.0   
65               73.0              36.0             37.0   
66               72.0              39.0             33.0   
67               78.0              37.0             41.0   
68              102.0              47.0             55.0   
352              70.0              35.0             35.0   
353              68.0              32.0             36.0   
354              70.0              32.0             38.0   
355              70.0              35.0             35.0   
356              70.0              35.0             35.0   

     Indigenous Enrolments (%)  Language Background Other Than English (%)  
64                         NaN                                         0.0  
65                         NaN                                         0.0  
66                         NaN                                         0.0  
67                         NaN                                         0.0  
68                         Na

In [56]:
len(indigenous_missing)

3067

It would appear that the schools are not reporting any indigenous enrollments.  It is hard as they may have enrolments but choose not to report them, in this case we will just fill the NaNs with 0 values. 

In [57]:
df5[['Indigenous Enrolments (%)']] = df5[['Indigenous Enrolments (%)']].fillna(0)

df5.isna().sum()

Calendar Year                                    0
School Name                                      0
Suburb                                           0
State                                            0
Postcode                                         0
School Sector                                    0
School Type                                      0
Campus Type                                      0
School URL                                       0
Governing Body                                   0
Geolocation                                      0
ICSEA                                            0
Teaching Staff                                   0
Non-Teaching Staff                               0
Total Enrolments                                 0
Girls Enrolments                                 0
Boys Enrolments                                  0
Indigenous Enrolments (%)                        0
Language Background Other Than English (%)    9307
dtype: int64

Lastly let us look at the Language Background Other than English.  This according to the dataset denotes the number of spoken languages by a school other than english, measured in a percent value. 

Let's check it out now

In [58]:
language_missing = df5[df5['Language Background Other Than English (%)'].isna()]
language_missing.head(10)

Calendar Year                    School Name               Suburb State  \
72             2008      Larmenier Catholic School          St Leonards   TAS   
204            2008  St Cuthbert's Catholic School          Lindisfarne   TAS   
276            2008   St Therese's Catholic School               Moonah   TAS   
420            2008   John Paul II Catholic School       Clarendon Vale   TAS   
2433           2008     Devonport Christian School                  Don   TAS   
2672           2008     Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2673           2009     Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2674           2010     Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2675           2011     Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   
2676           2012     Ngaanyatjarra Lands School  Ngaanyatjarra-Giles    WA   

      Postcode School Sector School Type           Campus Type  \
72        7250      Catholic     Primary  School Single Entity   
204       7015      Catholic     Primary  School Single Entity   
276       7009      Catholic     Primary  School Single Entity   
420       7019      Catholic     Primary  School Single Entity   
2433      7310   Independent     Primary  School Single Entity   
2672       872    Government    Combined    School Head Campus   
2673       872    Government    Combined    School Head Campus   
2674       872    Government    Combined    School Head Campus   
2675       872    Government    Combined    School Head Campus   
2676       872    Government    Combined    School Head Campus   

                             School URL  \
72    https://www.larmenier.tas.edu.au/   
204   http://www.stcuthberts.tas.edu.au   
276     http://www.sttherese.tas.edu.au   
420      http://www.johnpaul.tas.edu.au   
2433          http://www.dcs.tas.edu.au   
2672  http://www.nglandschool.wa.edu.au   
2673  http://www.nglandschool.wa.edu.au   
2674  http://www.nglandschool.wa.edu.au   
2675  http://www.nglandschool.wa.edu.au   
2676  http://www.nglandschool.wa.edu.au   

                                 Governing Body     Geolocation   ICSEA  \
72          Tasmanian Catholic Education Office  Inner Regional  1019.0   
204         Tasmanian Catholic Education Office  Inner Regional  1028.0   
276         Tasmanian Catholic Education Office  Inner Regional   982.0   
420         Tasmanian Catholic Education Office  Inner Regional   957.0   
2433               Independent Schools Tasmania  Inner Regional  1002.0   
2672  Department of Education Western Australia     Very Remote   510.0   
2673  Department of Education Western Australia     Very Remote   510.0   
2674  Department of Education Western Australia     Very Remote   510.0   
2675  Department of Education Western Australia     Very Remote   510.0   
2676  Department of Education Western Australia     Very Remote   570.0   

      Teaching Staff  Non-Teaching Staff  Total Enrolments  Girls Enrolments  \
72              19.0                 0.0             192.0              85.0   
204             16.0                10.0             205.0              91.0   
276             16.0                12.0             207.0             100.0   
420             10.0                 8.0             123.0              65.0   
2433             6.0                 8.0              87.0              46.0   
2672            26.0                28.0             364.0             177.0   
2673            36.0                22.0             345.0             172.0   
2674            30.0                29.0             335.0             166.0   
2675            42.0                19.0             310.0             144.0   
2676            48.0                 8.0             293.0             146.0   

      Boys Enrolments  Indigenous Enrolments (%)  \
72              107.0                        2.0   
204             114.0                        3.0   
276             107.0                        3.0  

Looking at this snip of information, it appears we are in the same boat as the Indigenous enrolments.  There are some schools who do not report the additional spoken languages than english however there is evidence they would have some sort of value.  As above let us set these at 0% as this is not detrimental to the end goal. 

In [59]:
df5[['Language Background Other Than English (%)']] = df5[['Language Background Other Than English (%)']].fillna(0)

df5.isna().sum()

Calendar Year                                 0
School Name                                   0
Suburb                                        0
State                                         0
Postcode                                      0
School Sector                                 0
School Type                                   0
Campus Type                                   0
School URL                                    0
Governing Body                                0
Geolocation                                   0
ICSEA                                         0
Teaching Staff                                0
Non-Teaching Staff                            0
Total Enrolments                              0
Girls Enrolments                              0
Boys Enrolments                               0
Indigenous Enrolments (%)                     0
Language Background Other Than English (%)    0
dtype: int64

In [60]:
# Lets save the cleaned file

df_ready = df5.reset_index(drop=True).copy()
df_ready.to_csv('apsev_clean_dataset.csv', index=False)

os.listdir(os.getcwd())

['.ipynb_checkpoints',
 'apsev_clean_dataset.csv',
 'APSEV_data_cleaning.ipynb',
 'school_profile_2008-2019.csv']

And we are Clean!!!

On to the Analysis!